### PR TITLE
os/arch/arm/src/amebad: Fix I2S for RTL8721CSM

### PR DESCRIPTION
**Pull request by jwei5**
os/arch/arm/src/amebad: Fix I2S for RTL8721CSM
- Add I2S and I2C options into Kconfig
- Fix I2S Tx and add 16bit padding feature for I2S (Comment #define __16_BIT_PADDING__ to disable feature)
- Note: Pin configs are unchanged

### CODE REVIEW

File: os/arch/arm/src/amebad/Kconfig [hunk 1]
```diff
@@ -51,6 +51,49 @@ config RTL8721D_SERIAL_FIFO
 	bool "UART SERIAL FIFO"
 	default n
 
+config AMEBAD_I2S
```
Comment (sunghan-chang) [inline os/arch/arm/src/amebad/Kconfig:54]: I think you should use this in Make.defs.
```
ifeq ($(CONFIG_AMEBAD_I2S),y)
CHIP_CSRCS += amebad_i2s.c
endif
```

File: os/arch/arm/src/amebad/amebad_i2s.c [hunk 2]
```diff
@@ -116,10 +117,13 @@
...
 #define OVER_SAMPLE_RATE (384U)
 
-#define I2S_DMA_PAGE_SIZE	768   // 2 ~ 4096
-#define I2S_DMA_PAGE_NUM    4   // Vaild number is 2~4
+#define I2S_DMA_PAGE_SIZE	1024   /* 4 ~ 16384, set to a factor of APB size */
+#define I2S_DMA_PAGE_NUM    4   /* Vaild number is 2~4 */
 
 /* I2S buffer container */
+#ifdef __16_BIT_PADDING__
+uint8_t audio_buf[I2S_DMA_PAGE_SIZE];
```
Comment (sunghan-chang) [inline os/arch/arm/src/amebad/amebad_i2s.c:125]: If this is used only in this file internally, let's add `static`.

File: os/arch/arm/src/amebad/amebad_i2s.c [hunk 3]
```diff
@@ -1056,7 +1127,7 @@ void i2s_transfer_rx_handleirq(void *data, char *pbuf)
 
 	int result = OK;
 
-	i2s_rxdma_callback(priv, result);
+	//i2s_rxdma_callback(priv, result);
```
Comment (sunghan-chang) [inline os/arch/arm/src/amebad/amebad_i2s.c:1130]: Do you need commented codes? I find several commented codes.

File: os/arch/arm/src/amebad/Kconfig (hunk 1 above)
Comment (jwei5) [reply os/arch/arm/src/amebad/Kconfig:54]: Ok, have added the following into Make.defs to compile amebad_i2s.c only when AMEBAD_I2S is configured

File: os/arch/arm/src/amebad/amebad_i2s.c (hunk 2 above)
Comment (jwei5) [reply os/arch/arm/src/amebad/amebad_i2s.c]: Yes it is only used within amebad_i2s.c, have added static to this variable

File: os/arch/arm/src/amebad/amebad_i2s.c (hunk 3 above)
Comment (jwei5) [reply os/arch/arm/src/amebad/amebad_i2s.c]: Ok, I have removed the commented codes that are not needed, such as i2s_txdma_prep()

File: os/arch/arm/src/amebad/amebad_i2s.c [hunk 4]
```diff
@@ -116,10 +117,13 @@
 
 #define OVER_SAMPLE_RATE (384U)
 
-#define I2S_DMA_PAGE_SIZE	768   // 2 ~ 4096
-#define I2S_DMA_PAGE_NUM    4   // Vaild number is 2~4
+#define I2S_DMA_PAGE_SIZE	1024   /* 4 ~ 16384, set to a factor of APB size */
+#define I2S_DMA_PAGE_NUM    4   /* Vaild number is 2~4 */
 
 /* I2S buffer container */
+#ifdef __16_BIT_PADDING__
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/amebad/amebad_i2s.c:124]: above default values are acceptable, but regarding PADDING, we need a KConfig

**Review by Taejun-Kwon — COMMENTED** [review]
@abhishek-samsung could you please compare source code with our latest?

File: os/arch/arm/src/amebad/amebad_i2s.c (hunk 4 above)
Comment (jwei5) [reply os/arch/arm/src/amebad/amebad_i2s.c]: Noted, I have added a AMEBAD_I2S_16BIT_PADDING option in KConfig to turn on the feature

**Comment by abhishek-samsung** [comment]
> @abhishek-samsung could you please compare source code with our latest?

Hi @Taejun-Kwon , I have verified, only this line has been commented our latest but is not commented out in this PR.

https://github.com/abhishek-samsung/TizenRT/blob/387b607fb174b262a56405bfa8b5746878499184/os/arch/arm/src/amebad/amebad_i2s.c#L1130

amebad_i2s.c line 1116 in this PR.

@jwei5 , could you please check this?

**Comment by jwei5** [comment]
> 

Hi @abhishek-samsung , the uncommented line for i2s_rxdma_callback(priv, result) is not part of the changes for this commit. I checked that the line was commented in one of the commits in the ub6470 branch, I will comment it here as well to sync with the source code.

File: os/arch/arm/src/amebad/amebad_i2s.c [hunk 5]
```diff
@@ -1056,7 +1113,7 @@ void i2s_transfer_rx_handleirq(void *data, char *pbuf)
 
 	int result = OK;
 
-	i2s_rxdma_callback(priv, result);
+	// i2s_rxdma_callback(priv, result);
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/amebad/amebad_i2s.c:1116]: We should not disable this if we consider rx in future, how do you think?

Comment (jwei5) [reply os/arch/arm/src/amebad/amebad_i2s.c:1116]: Hi @abhishek-samsung , could I check with you on the reason this line is commented? I see that it had been commented out during this [commit](https://github.com/abhishek-samsung/TizenRT/commit/2d164c96ddfbb916bdbf08e7a3aa71dc7278067b#diff-174f6cd168a26967c6eb71d3acb91b4e3471b6997ec3e05aa762de877c87ddb0).

Comment (abhishek-samsung) [reply os/arch/arm/src/amebad/amebad_i2s.c:1116]: Hi @jwei5 , earlier we were not using i2s rx, so I have commented out the line. Please uncomment the line. Sorry for the confusion.

Comment (jwei5) [reply os/arch/arm/src/amebad/amebad_i2s.c]: Ok no problem, thanks for the information, I have uncommented the line.

File: os/arch/arm/src/amebad/Kconfig [hunk 6]
```diff
@@ -51,6 +51,55 @@ config RTL8721D_SERIAL_FIFO
...
+	---help---
+		Enable 16-bit padding to send 16bit I2S data in 32bit word length
+endif
+
+config AMEBAD_I2C_DYNTIMEO
+	bool "[Amebad I2C]Use dynamic timeouts"
+	default n
+
+config AMEBAD_I2C_DYNTIMEO_USECPERBYTE
+	int "[Amebad I2C]Timeout Microseconds per Byte"
```
Comment (kishore-sn) [inline os/arch/arm/src/amebad/Kconfig:79]: Is it necessary to mention `[Amebad I2C]` for every config option?

Comment (jwei5) [reply os/arch/arm/src/amebad/Kconfig:79]: Hi, originally it was meant to indicate that the option is for I2C
How about removing `[Amebad I2C]` and adding `comment "I2C Options"` above the I2C configs, will that be better?

**Review by Taejun-Kwon — APPROVED** [review]

**Pull request merged**